Automattic/vip-cli
# Flow type checking is broken 😢

## Description

While working in this repo, I encountered some really strange behavior where Flow errors would sometimes surface locally but disappear on subsequent runs—or appear in WebStorm but not in Vim. I shrugged them off because Flow type checks always seemed to pass in CI. However, this was a mirage: 🏝️ 

```sh
$ npm run flow

> @automattic/vip@2.0.14 flow /Users/czarate/Documents/Code/vip
> flow

Launching Flow server for /Users/czarate/Documents/Code/vip
Spawned flow server (pid=24473)
Logs will go to /private/tmp/flow/zSUserszSczZaratezSDocumentszSCodezSvip.log
Monitor logs will go to /private/tmp/flow/zSUserszSczZaratezSDocumentszSCodezSvip.monitor_log
No errors!
```

The output says no errors, but it notes that logs are sent to a temp file. If we inspect that file, we can see that all is not well:

```
[snipped]
[2021-09-10 13:08:07.053] Persistent response: liveErrorsResponse OK (3 errors, 0 warnings, 0 other) file:///Users/czarate/Documents/Code/vip/src/lib/validations/site-type.js
[2021-09-10 13:08:07.077] Persistent response: liveErrorsResponse OK (2 errors, 0 warnings, 0 other) file:///Users/czarate/Documents/Code/vip/src/lib/validations/sql.js
```

It appears the culprit is our use of `resolve_dirname` in `.flowconfig`:

https://github.com/facebook/flow/issues/5180#issuecomment-341887514

We are essentially telling Flow to treat those files as `node_modules`, which means the type checking is skipped in many contexts. This is not a bug, but intended behavior.

The fix is to move to `module.name_mapper` instead, which shakes loose an avalanche of type errors. I'm happy to work on them in separate PRs but wanted to check to make sure effort would not be better spent migrating to TypeScript. (Good news: Our other projects that use Flow do not use `resolve_dirname`.)

## Steps to Test

1. Check out PR.
1. Run `npm run flow`
1. 😭
